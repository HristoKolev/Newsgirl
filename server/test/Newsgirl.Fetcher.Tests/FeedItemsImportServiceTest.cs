using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

using NSubstitute;
using Xunit;
using Newsgirl.Testing;

namespace Newsgirl.Fetcher.Tests
{
    using Shared;

    public class FeedItemsImportServiceTest : DatabaseTest
    {
        [Fact]
        public async Task GetFeedsForUpdate_Returns_All_Feeds()
        {
            var log = new StructuredLogMock();

            var importService = new FeedItemsImportService(this.Db, DbConnection, log);

            var feeds = Enumerable.Range(1, 10)
                .Select(i => new FeedPoco
                {
                    FeedItemsHash = i,
                    FeedContentHash = i,
                    FeedName = $"feed {i}",
                    FeedUrl = $"url {i}",
                }).ToList();

            await this.Db.BulkInsert(feeds);
            
            var resultFeeds = await importService.GetFeedsForUpdate();
            
            Snapshot.Match(resultFeeds);
        }
        
        
        [Fact]
        public async Task ImportItems_Copies_Items_Correctly()
        {
            var log = new StructuredLogMock();

            var importService = new FeedItemsImportService(this.Db, this.DbConnection, log);

            var feeds = Enumerable.Range(1, 10)
                .Select(i => new FeedPoco
                {
                    FeedName = $"feed {i}",
                    FeedUrl = $"url {i}",
                }).ToList();

            foreach (var feed in feeds)
            {
                await this.Db.Insert(feed);
            }

            var updates = new List<FeedUpdateModel>
            {
                new FeedUpdateModel
                {
                    NewItems = Enumerable.Range(1, 10)
                        .Select(i => new FeedItemPoco
                        {
                            FeedID = 1,
                            FeedItemDescription = $"desc {i}",
                            FeedItemHash = i,
                            FeedItemTitle = $"title {i}",
                            FeedItemUrl = $"url {i}",
                            FeedItemAddedTime = TestHelper.Date2000,
                        }).ToList(),
                    Feed = feeds.First(x => x.FeedID == 1),
                    NewFeedItemsHash = 1,
                },
                new FeedUpdateModel
                {
                    NewItems = Enumerable.Range(100, 10)
                        .Select(i => new FeedItemPoco
                        {
                            FeedID = 2,
                            FeedItemDescription = $"desc {100 + i}",
                            FeedItemHash = i,
                            FeedItemTitle = $"title {100 + i}",
                            FeedItemUrl = $"url {100 + i}",
                            FeedItemAddedTime = TestHelper.Date2000,
                        }).ToList(),
                    Feed = feeds.First(x => x.FeedID == 2),
                    NewFeedItemsHash = 2,
                },
                new FeedUpdateModel
                {
                    Feed = feeds.First(x => x.FeedID == 3), 
                },
                new FeedUpdateModel
                {
                    Feed = feeds.First(x => x.FeedID == 4),
                    NewItems = new List<FeedItemPoco>(),
                    NewFeedItemsHash = 4
                },
                new FeedUpdateModel
                {
                    NewItems = Enumerable.Range(200, 10)
                        .Select(i => new FeedItemPoco
                        {
                            FeedID = 5,
                            FeedItemDescription = null,
                            FeedItemHash = i,
                            FeedItemTitle = $"title {i}",
                            FeedItemUrl = null,
                            FeedItemAddedTime = TestHelper.Date2000,
                        }).ToList(),
                    Feed = feeds.First(x => x.FeedID == 5),
                    NewFeedItemsHash = 5,
                },
            };
            
            await importService.ImportItems(updates.ToArray());

            var resultFeeds = this.Db.Poco.Feeds.OrderByDescending(x => x.FeedID).ToList();
            var resultFeedItems = this.Db.Poco.FeedItems.OrderByDescending(x => x.FeedItemID).ToList();
            
            Snapshot.Match(new {resultFeeds, resultFeedItems});
        }
        
        [Fact]
        public async Task GetMissingFeedItems_Returns_Correct_Result()
        {
            var log = new StructuredLogMock();

            var importService = new FeedItemsImportService(this.Db, this.DbConnection, log);

            var feeds = Enumerable.Range(1, 10)
                .Select(i => new FeedPoco
                {
                    FeedName = $"feed {i}",
                    FeedUrl = $"url {i}",
                }).ToList();

            await this.Db.BulkInsert(feeds);

            var feedItems = Enumerable.Range(1, 10)
                .Select(i => new FeedItemPoco
                {
                    FeedID = 1,
                    FeedItemHash = i,
                    FeedItemDescription = $"desc {i}",
                    FeedItemTitle = $"title {i}",
                    FeedItemUrl = $"url {i}",
                    FeedItemAddedTime = TestHelper.Date2000,
                }).ToList();
            
            feedItems.AddRange(Enumerable.Range(30, 10)
                .Select(i => new FeedItemPoco
                {
                    FeedID = 2,
                    FeedItemHash = i,
                    FeedItemDescription = $"desc {i}",
                    FeedItemTitle = $"title {i}",
                    FeedItemUrl = $"url {i}",
                    FeedItemAddedTime = TestHelper.Date2000,
                }));

            await this.Db.BulkInsert(feedItems);

            long[] hashes = {
                1, 2, 3, 31, 32, 33
            };
            
            var missingFeedItems = await importService.GetMissingFeedItems(1, hashes);
            
            Array.Sort(missingFeedItems);
                        
            Snapshot.Match(missingFeedItems);
        }
    }
}
