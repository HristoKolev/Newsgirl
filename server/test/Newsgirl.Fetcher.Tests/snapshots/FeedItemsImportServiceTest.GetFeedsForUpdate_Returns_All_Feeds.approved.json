[
  {
    "FeedHash": 10,
    "FeedID": 10,
    "FeedName": "feed 10",
    "FeedUrl": "url 10"
  },
  {
    "FeedHash": 9,
    "FeedID": 9,
    "FeedName": "feed 9",
    "FeedUrl": "url 9"
  },
  {
    "FeedHash": 8,
    "FeedID": 8,
    "FeedName": "feed 8",
    "FeedUrl": "url 8"
  },
  {
    "FeedHash": 7,
    "FeedID": 7,
    "FeedName": "feed 7",
    "FeedUrl": "url 7"
  },
  {
    "FeedHash": 6,
    "FeedID": 6,
    "FeedName": "feed 6",
    "FeedUrl": "url 6"
  },
  {
    "FeedHash": 5,
    "FeedID": 5,
    "FeedName": "feed 5",
    "FeedUrl": "url 5"
  },
  {
    "FeedHash": 4,
    "FeedID": 4,
    "FeedName": "feed 4",
    "FeedUrl": "url 4"
  },
  {
    "FeedHash": 3,
    "FeedID": 3,
    "FeedName": "feed 3",
    "FeedUrl": "url 3"
  },
  {
    "FeedHash": 2,
    "FeedID": 2,
    "FeedName": "feed 2",
    "FeedUrl": "url 2"
  },
  {
    "FeedHash": 1,
    "FeedID": 1,
    "FeedName": "feed 1",
    "FeedUrl": "url 1"
  }
]